### PR TITLE
Use async requests to delete runs and add view to get runs by run name more efficiently

### PR DESCRIPTION
**Pull request by eamansour**
## Why?
See https://github.com/galasa-dev/projectmanagement/issues/2020

## Changes
- Deleting all associated documents of a run is now done asynchronously to speed up deletion
- Added a new run name view to get runs by run name (which is faster than using a mango query)
  - Changed the `GET /ras/runs` route to use this new view when getting runs by run name

Local testing shows an improvement in both `runs get --name xxx` and `runs delete`.

Previously:
```
❯ time galasactl runs get --name C11523
submitted-time(UTC) name   requestor              status   result test-name
2024-09-27 15:30:20 C11523 galasadelivery@ibm.com finished Passed dev.galasa.inttests.core.local.CoreLocalJava11Ubuntu

Total:1 Passed:1
galasactl runs get --name C11523  0.00s user 0.01s system 0% cpu 1.734 total

❯ time galasactl runs delete --name C11523
galasactl runs delete --name C11523  0.00s user 0.01s system 0% cpu 2.002 total
```

Now:
```
❯ time galasactl runs get --name C11522
submitted-time(UTC) name   requestor              status   result test-name
2024-09-27 15:19:50 C11522 galasadelivery@ibm.com finished Passed dev.galasa.inttests.core.local.CoreLocalJava11Ubuntu

Total:1 Passed:1
galasactl runs get --name C11522  0.00s user 0.01s system 14% cpu 0.091 total

❯ time galasactl runs delete --name C11522
galasactl runs delete --name C11522  0.00s user 0.00s system 7% cpu 0.109 total
```

### CODE REVIEW

File: modules/extensions/galasa-extensions-parent/dev.galasa.extensions.common/src/main/java/dev/galasa/extensions/common/couchdb/CouchdbStore.java [hunk 1]
```diff
@@ -147,13 +148,43 @@ protected List<ViewRow> getAllDocsFromDatabase(String dbName) throws CouchdbExce
...
+            if (viewResponse.rows == null) {
+                String errorMessage = ERROR_FAILED_TO_GET_VIEW_DOCUMENTS_FROM_DATABASE.getMessage(viewName, dbName);
+                throw new CouchdbException(errorMessage);
+            }
+        } catch (URISyntaxException e) {
+            String errorMessage = ERROR_URI_IS_INVALID.getMessage(viewRequestUriStr);
+            throw new CouchdbException(errorMessage, e);
+        }
+        return viewResponse;
+    }
```
Comment (techcobweb) [inline modules/extensions/galasa-extensions-parent/dev.galasa.extensions.common/src/main/java/dev/galasa/extensions/common/couchdb/CouchdbStore.java:186]: Wondering if the couchdb part of the extensions/common should be in a separate ...common.couchdb bundle. As the etcd extensions don't really need this code ?

File: modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/ras/directory/DirectoryRASDirectoryService.java [hunk 2]
```diff
@@ -70,10 +70,24 @@ protected DirectoryRASDirectoryService(@NotNull Path baseDirectory, GalasaGson g
         return runs;
     }
 
+
+    @Override
+    public List<IRunResult> getRunsByRunName(@NotNull String runName) throws ResultArchiveStoreException {
+        List<DirectoryRASRunResult> allRuns = getAllRuns();
```
Comment (techcobweb) [inline modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/ras/directory/DirectoryRASDirectoryService.java:76]: Does this require all the runs in the entire database ?

File: modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java [hunk 3]
```diff
@@ -432,76 +469,58 @@ private JsonArray buildQuerySortJson(@NotNull RasSortField primarySort) {
...
-
-            discardRecord(CouchdbRasStore.RUNS_DB, runTestStructure._id, runTestStructure._rev);
-        } catch (CouchdbRasException | ParseException e) {
+            // Build a list of discard operation futures
+            List<CompletableFuture<Void>> futures = discardRecords(LOG_DB, runTestStructure.getLogRecordIds());
+            futures.addAll(discardRecords(ARTIFACTS_DB, runTestStructure.getArtifactRecordIds()));
+            futures.add(discardRecord(RUNS_DB, runTestStructure._id, runTestStructure._rev));
+
+            // Wait for all the discard operations to finish
+            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
```
Comment (techcobweb) [inline modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java]: Can any of those futures block ? I assume they timeout or something ?

File: modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java [hunk 4]
```diff
@@ -432,76 +469,58 @@ private JsonArray buildQuerySortJson(@NotNull RasSortField primarySort) {
...
-                throw new CouchdbRasException("Unable to find runs - " + statusLine.toString());
-            }
-
-            IdRev found = store.getGson().fromJson(responseEntity, IdRev.class);
-            if (found._id == null) {
-                throw new CouchdbRasException("Unable to find runs - Invalid JSON response");
-            }
-            if (found._rev == null) {
-                throw new CouchdbRasException("Unable to find rev - Invalid JSON response");
+    private CompletableFuture<String> getRevision(String databaseName, String id) {
```
Comment (techcobweb) [inline modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java]: I might have been tempted to separate the async/future stuff into a separate class. This class is getting huge. Or even just separate out the deletion implementation as it's kinda separate/doing things in a separate way to the rest ?

**Review by techcobweb — APPROVED** [review]
Some comments/questions we can talk about.

File: modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/ras/directory/DirectoryRASDirectoryService.java (hunk 2 above)
Comment (techcobweb) [reply modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/ras/directory/DirectoryRASDirectoryService.java:76]: This is the local implementation so we are not dealing with couchdb... ok

File: modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java (hunk 3 above)
Comment (techcobweb) [reply modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java]: Yes. We assume that the individual requests would time out.

File: modules/extensions/galasa-extensions-parent/dev.galasa.extensions.common/src/main/java/dev/galasa/extensions/common/couchdb/CouchdbStore.java (hunk 1 above)
Comment (eamansour) [reply modules/extensions/galasa-extensions-parent/dev.galasa.extensions.common/src/main/java/dev/galasa/extensions/common/couchdb/CouchdbStore.java:186]: Thanks, added `dev.galasa.extensions.common.couchdb` bundle in https://github.com/galasa-dev/galasa/pull/108/commits/adc238a09a558157c79900303f414bc140af090a

File: modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java (hunk 4 above)
Comment (eamansour) [reply modules/extensions/galasa-extensions-parent/dev.galasa.ras.couchdb/src/main/java/dev/galasa/ras/couchdb/internal/CouchdbDirectoryService.java]: Agreed - moved the deletion code into a separate class in https://github.com/galasa-dev/galasa/pull/108/commits/fa9d1a5443776870efa5b5cf710205d2e2999fcb

**Pull request merged**